File: Regression/Section 6 - Polynomial Regression/Polynomial_Regression/trueSalaryPredictor.R
# read data
dataSet = dataSet[, 2:3]

#Insufficient data, Hence no need to split as train and test data

# Fit to polynomial Regresion
dataSet$degree2 = dataSet$Level^2
dataSet$degree3 = dataSet$Level^3
dataSet$degree4 = dataSet$Level^4

regressor = lm(formula = Salary ~ Level + degree2 + degree3 + degree4, data = dataSet)
# Visualize Result
library(ggplot2)
xGrid = seq(min(dataSet$Level), max(dataSet$Level), .1)
ggplot()+
  geom_point(aes(x=dataSet$Level, y=dataSet$Salary), colour = 'RED')+
  geom_line(aes(x=xGrid, y=predict(regressor, newdata = data.frame(Level = xGrid,
                                                                   degree2 = xGrid^2,
                                                                   degree3 = xGrid^3,
                                                                   degree4=xGrid^4))), colour = 'BLUE')+
  ggtitle('Employee Level VS Salary')+
  xlab('Level')+
  ylab('Salary')

# Predicitng salary of any level employee
yPredict = predict(regressor, newdata = data.frame(Level = x,
                                                   degree2 = x^2,
                                                   degree3 = x^3,
                                                   degree4 = x^4))

File: Regression/Section 6 - Polynomial Regression/Polynomial_Regression/trueSalaryPredictor.py
# import libraries
import pandas as pd
import numpy as nm
import matplotlib.pyplot as plt

# import data set
dataSet = pd.read_csv('Position_Salaries.csv')
x = dataSet.iloc[:, 1:2].values
y = dataSet.iloc[:, 2].values

# Insufficient data and hence no train test data split
# Fit to linear regression
from sklearn.linear_model import LinearRegression
reg1 = LinearRegression()
reg1.fit(x, y)

# Fit to Polynomial regression
from sklearn.preprocessing import PolynomialFeatures
polyReg = PolynomialFeatures(degree=4)
xPoly = polyReg.fit_transform(x)
reg2 = LinearRegression()
reg2.fit(xPoly, y)

#Visualizing the Linear regression result
plt.scatter(x, y, color = 'RED')
plt.plot(x, reg1.predict(x), color = 'BLUE')
plt.title('Employee Position VS Salary')
plt.xlabel('Position Level')
plt.ylabel('Salary')
plt.show()

""" The visualization graph shows huge deviation of the predicted 
salary from actual salary. Hence linear regression model isn't 
useful in this situation. Therefore, Polynomial model is used"""

# Visualize the polynomial result
#Visualizing the Linear regression result
xGrid = nm.arange(min(x), max(x), .1)
xGrid = xGrid.reshape(len(xGrid),1)
plt.scatter(x, y, color = 'RED')
plt.plot(x, reg2.predict(polyReg.fit_transform(x)), color = 'BLUE')
plt.title('Employee Position VS Salary')
plt.xlabel('Position Level')
plt.ylabel('Salary')
plt.show()

"""To predict the salry of a employee of any level,
substitute x with the actual level in the following line of code"""
reg2.predict(polyReg.fit_transform(x))